pkl-cli/src/main/kotlin/org/pkl/cli/CliEvaluator.kt

```diff
@@ -174,34 +196,49 @@ constructor(
                 StandardOpenOption.WRITE,
                 StandardOpenOption.APPEND,
               )
-              outputFile.writeString(
-                output,
-                Charsets.UTF_8,
-                StandardOpenOption.WRITE,
-                StandardOpenOption.APPEND,
-              )
+              outputFile.writeBytes(output, StandardOpenOption.WRITE, StandardOpenOption.APPEND)
             }
           }
         }
       } else {
         var outputWritten = false
         for (moduleUri in options.base.normalizedSourceModules) {
-          val moduleSource = toModuleSource(moduleUri, consoleReader)
-          val output = evaluator.evaluateExpressionString(moduleSource, options.expression)
-          if (output.isNotEmpty()) {
-            if (outputWritten) consoleWriter.appendLine(options.moduleOutputSeparator)
-            consoleWriter.write(output)
-            consoleWriter.flush()
-            outputWritten = true
+          val moduleSource = toModuleSource(moduleUri, inputStream)
+          if (options.expression != null) {
+            val output = evaluator.evaluateExpressionString(moduleSource, options.expression)
+            if (output.isNotEmpty()) {
+              if (outputWritten) outputStream.writeLine(options.moduleOutputSeparator)
+              outputStream.writeText(output)
+              outputStream.flush()
+              outputWritten = true
+            }
+          } else {
+            val outputBytes = evaluator.evaluateOutputBytes(moduleSource)
+            if (outputBytes.isNotEmpty()) {
+              if (outputWritten) outputStream.writeLine(options.moduleOutputSeparator)
+              outputStream.write(outputBytes)
+              outputStream.flush()
+              outputWritten = true
+            }
           }
         }
       }
     }
   }
 
-  private fun toModuleSource(uri: URI, reader: Reader) =
-    if (uri == VmUtils.REPL_TEXT_URI) ModuleSource.create(uri, reader.readText())
-    else ModuleSource.uri(uri)
+  private fun OutputStream.writeText(text: String) = write(text.toByteArray())
```

Comment:
Should we be using `toByteArray(StandardCharsets.UTF_8)` here?

Reply:
That's the default argument, but, fair, better to be clear here.